servers/cu/src/domain/utils.js

```diff
@@ -422,3 +424,73 @@ export const maybeParseInt = pipe(
   mapBadData,
   (val) => val ? parseInt(val) : undefined
 )
+
+export function isJsonString (str) {
+  try {
+    JSON.parse(str)
+    return true
+  } catch (e) {
+    return false
+  }
+}
+
+const SIG_CONFIG = {
+  ARWEAVE: {
+    sigLength: 512,
+    pubLength: 512,
+    sigName: 'arweave'
+  },
+  ETHEREUM: {
+    sigLength: 65,
+    pubLength: 65,
+    sigName: 'ethereum'
+  }
+}
+
+/**
+ * From etherium docs: https://ethereum.org/en/developers/docs/accounts/#account-creation
+ *
+ * You get a public address by taking the last 20 bytes of
+ * the Keccak-256 hash of the public key and adding 0x to the beginning.
+ */
+function keyToEtheriumAddress (key) {
+  const hash = createKeccakHash('keccak256')
+
+  hash.update(Buffer.from(key, 'base64url'))
+
+  const hashed = hash.digest('hex')
+  /**
+   * last 20 bytes of the hash, prefixed with '0x'
+   */
+  return '0x' + hashed.slice(-40)
+}
+
+const cache = new LRUCache({ max: 1000 })
+/**
+ * Given the address and public key, generate the owner.
+ * This function is memoized, and deduped, for performance, caching the most recent
+ * 1000 computed owners
+ *
+ * @param {{ address: string, key: string }} arg
+ * @returns {string}
+ */
+export const addressFrom = ({ address, key }) => {
+  const cKey = `${address}-${key}`
+  if (cache.has(cKey)) return cache.get(cKey)
+
+  const pubKeyLength = Buffer.from(key, 'base64url').length
+  let _address
+
+  if (pubKeyLength === SIG_CONFIG.ETHEREUM.pubLength) _address = keyToEtheriumAddress(key)
```

Comment:
Determining the signature type of a bundle item by the length of its public key is not a good way to do this.
With the ANS104 standard, the bundle item has a signature type field to mark the type of public key used for signing.
ANS104: https://github.com/joshbenaron/arweave-standards/blob/ans104/ans/ANS-104.md
So it is possible to determine the address type simply by signatur type

Reply:
<img width="460" alt="image" src="https://github.com/permaweb/ao/assets/40423181/53c506f5-af13-4398-a30a-3c8de95de3fc">
This situation would be impossible to judge

Reply:
You're right, and I agree, that using signature type would be ideal. 

The impediment is that the CU actually does not handle raw Data Items directly, but retrieves message/process metadata from either the SU scheduling the process, or an Arweave Gateway (in the case of assignments), both of which do not currently expose the signature type. So currently, the CU simply does not have the signature type handy to make that distinction. However, it does have the pub key.

Using pub key length isn't ideal, or foolproof when considering all signature types, but it does allow for distinguishing Arweave from Ethereum, which is the focus of this change. It of course doesn't work when other signature types with same key lengths are considered, but then the impl just behaves as it does now (see the TODO on the fall through case in `addressFrom`). In this sense, these changes are additive. 

Ideally, the SU and gateway schemas can be extended to expose signature type, at which point the CU can and ought to use that to make this distinction. Until then, this gets what we need.

Reply:
If this is the case then indeed the fastest and easiest way to integrate ethereum addresses is to use public length.
Hopefully this pr will be merged soon. I proposed support for ethereum address in an issue two days ago, and I didn't realise you were already implementing it. Thank you very much.
If this feature is merged into ao, the ao cross-chain bridge will natively support ethereum addresses, and users will be able to use the metamask to sign and query transactions and balances on ao, which is great for the whole ao ecosystem!